unittests/db_to_kv_tests.cpp

```diff
@@ -0,0 +1,309 @@
+                       "th character is actually greater on the left than the right.  Called from: " + desc);
+   BOOST_CHECK_MESSAGE(ret.first != 0, "expected " + std::to_string(lhs_index) +
+                       "th composite key to be less than the " + std::to_string(lhs_index + 1) +
+                       "th, but they were identical.  Called from: " + desc);
+}
+
+template<typename T, typename KeyFunc>
+void verify_scope_table_order(T low_key, T high_key, KeyFunc key_func, bool skip_trailing_prim_key) {
+   std::vector<b1::chain_kv::bytes> comp_keys;
+   comp_keys.push_back(key_func(name{0}, name{0}, low_key, 0x0));
+   // this function always passes the extra primary key, but need to skip what would be redundant keys for when we are
+   // actually getting a primary key
+   if (!skip_trailing_prim_key) {
+      comp_keys.push_back(key_func(name{0}, name{0}, low_key, 0x1));
+      comp_keys.push_back(key_func(name{0}, name{0}, low_key, 0xffff'ffff'ffff'ffff));
+   }
+   comp_keys.push_back(key_func(name{1}, name{0}, low_key, 0x0));
+   comp_keys.push_back(key_func(name{1}, name{1}, low_key, 0x0));
+   comp_keys.push_back(key_func(name{1}, name{1}, high_key, 0x0));
+   if (!skip_trailing_prim_key) {
+      comp_keys.push_back(key_func(name{1}, name{1}, high_key, 0x1));
+   }
+   comp_keys.push_back(key_func(name{1}, name{2}, low_key, 0x0));
+   comp_keys.push_back(key_func(name{1}, name{0xffff'ffff'ffff'ffff}, low_key, 0x0));
+   comp_keys.push_back(key_func(name{0xffff'ffff'ffff'ffff}, name{0}, low_key, 0x0));
+   comp_keys.push_back(key_func(name{0xffff'ffff'ffff'ffff}, name{0xffff'ffff'ffff'ffff}, low_key, 0x0));
+   for (unsigned int i = 0; i < comp_keys.size() - 1; ++i) {
+      verify_assending_composite_keys(comp_keys, i, "verify_scope_table_order");
+   }
+}
+
+// using this function to allow check_ordered_keys to work for all key types
+b1::chain_kv::bytes prim_drop_extra_key(name scope, name table, uint64_t db_key, uint64_t unused_primary_key) {
+   return eosio::chain::db_key_value_format::create_primary_key(scope, table, db_key);
+};
+
+template<typename T, typename KeyFunc>
+void check_ordered_keys(const std::vector<T> ordered_keys, KeyFunc key_func, bool skip_trailing_prim_key = false) {
+   verify_scope_table_order(ordered_keys[0], ordered_keys[1], key_func, skip_trailing_prim_key);
+   name scope{0};
+   name table{0};
+   const uint64_t prim_key = 0;
+   std::vector<b1::chain_kv::bytes> composite_keys;
+   for (const auto key: ordered_keys) {
+      auto comp_key = key_func(scope, table, key, prim_key);
+      composite_keys.push_back(comp_key);
+   }
+   for (unsigned int i = 0; i < composite_keys.size() - 1; ++i) {
+      verify_assending_composite_keys(composite_keys, i, "check_ordered_keys");
+   }
+}
+
+BOOST_AUTO_TEST_CASE(ui64_keys_conversions_test) {
+   const name scope = N(myscope);
```

Comment:
Could the following three variables be constexpr?

Reply:
I think we could, but I think here it is cleaner to read to just be a const, since this is test code.